ctypesgen/printer_python/preamble.py

```diff
@@ -16,364 +15,25 @@
-        return self.__class__(self.data.zfill(width))
-
-
-class MutableString(UserString):
-    """mutable string objects
-
-    Python strings are immutable objects.  This has the advantage, that
-    strings may be used as dictionary keys.  If this property isn't needed
-    and you insist on changing string values in place instead, you may cheat
-    and use MutableString.
-
-    But the purpose of this class is an educational one: to prevent
-    people from inventing their own mutable string class derived
-    from UserString and than forget thereby to remove (override) the
-    __hash__ method inherited from UserString.  This would lead to
-    errors that would be very hard to track down.
-
-    A faster and better solution is to rewrite your program using lists."""
-
-    def __init__(self, string=""):
-        self.data = string
-
-    def __hash__(self):
-        raise TypeError("unhashable type (it is mutable)")
-
-    def __setitem__(self, index, sub):
-        if index < 0:
-            index += len(self.data)
-        if index < 0 or index >= len(self.data):
-            raise IndexError
-        self.data = self.data[:index] + sub + self.data[index + 1 :]
-
-    def __delitem__(self, index):
-        if index < 0:
-            index += len(self.data)
-        if index < 0 or index >= len(self.data):
-            raise IndexError
-        self.data = self.data[:index] + self.data[index + 1 :]
-
-    def __setslice__(self, start, end, sub):
-        start = max(start, 0)
-        end = max(end, 0)
-        if isinstance(sub, UserString):
-            self.data = self.data[:start] + sub.data + self.data[end:]
-        elif isinstance(sub, bytes):
-            self.data = self.data[:start] + sub + self.data[end:]
-        else:
-            self.data = self.data[:start] + str(sub).encode() + self.data[end:]
-
-    def __delslice__(self, start, end):
-        start = max(start, 0)
-        end = max(end, 0)
-        self.data = self.data[:start] + self.data[end:]
-
-    def immutable(self):
-        return UserString(self.data)
-
-    def __iadd__(self, other):
-        if isinstance(other, UserString):
-            self.data += other.data
-        elif isinstance(other, bytes):
-            self.data += other
-        else:
-            self.data += str(other).encode()
-        return self
-
-    def __imul__(self, n):
-        self.data *= n
-        return self
-
-
-class String(MutableString, ctypes.Union):
-    _fields_ = [("raw", ctypes.POINTER(ctypes.c_char)), ("data", ctypes.c_char_p)]
-
-    def __init__(self, obj=b""):
-        if isinstance(obj, (bytes, UserString)):
-            self.data = bytes(obj)
-        else:
-            self.raw = obj
-
-    def __len__(self):
-        return self.data and len(self.data) or 0
+class String(ctypes.c_char_p):
+    @classmethod
+    def _check_retval_(cls, result):
+        value = result.value
+        return value if value is None else _bytes_wrapper(value)
```

Comment:
I guess this isn't ideal since the caller looses access to the original `c_char_p`.
Maybe it would be better to return a wrapper class around the raw `c_char_p` instead, with `__getattr__()` to the value, if that works?

Reply:
I tried this locally, but then methods like `upper()` return new bytes objects, which is kind of confusing.
I pushed it for the record, though. Did the original string classes behave like this, too? It looks like it...